data_model/src/events/data/filters.rs

```diff
@@ -0,0 +1,285 @@
+pub type OtherDomainChangeFilter = detail::SimpleEntityFilter<DomainId>;
+pub type OtherAccountChangeFilter = detail::SimpleEntityFilter<AccountId>;
+pub type DomainEntityFilter = detail::ComplexEntityFilter<DomainEvent, DomainEventFilter>;
+pub type AccountEntityFilter = detail::ComplexEntityFilter<AccountEvent, AccountEventFilter>;
+pub type EventFilter = FilterOpt<EntityFilter>;
+
+mod detail {
+    //! This module contains *sealed* structs, that is used in public API, but should
+    //! not be accessed from nowhere except parent module
+
+    use super::*;
+
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub struct SimpleEntityFilter<Id: Eq> {
+        id_filter: IdFilter<Id>,
+        status_filter: StatusFilter,
+    }
+
+    impl<Id: Eq> SimpleEntityFilter<Id> {
+        pub fn new(id_filter: IdFilter<Id>, status_filter: StatusFilter) -> Self {
+            Self {
+                id_filter,
+                status_filter,
+            }
+        }
+    }
+
+    impl<Id: Into<IdBox> + Debug + Clone + Eq + Ord> Filter for SimpleEntityFilter<Id> {
+        type Item = SimpleEvent<Id>;
+
+        fn filter(&self, entity: &SimpleEvent<Id>) -> bool {
+            self.id_filter.filter(entity.id()) && self.status_filter.filter(entity.status())
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub struct ComplexEntityFilter<Entity, EventFilter>
+    where
+        Entity: IdTrait,
+        EventFilter: Filter<Item = Entity>,
+        for<'a> <Entity as Identifiable>::Id: IntoSchema + Deserialize<'a> + Serialize,
+    {
+        id_filter: FilterOpt<IdFilter<<Entity as Identifiable>::Id>>,
+        event_filter: FilterOpt<EventFilter>,
+    }
+
+    impl<Entity, EventFilter> ComplexEntityFilter<Entity, EventFilter>
+    where
+        Entity: IdTrait,
+        EventFilter: Filter<Item = Entity>,
+        for<'a> <Entity as Identifiable>::Id: IntoSchema + Deserialize<'a> + Serialize,
+    {
+        pub fn new(
+            id_filter: FilterOpt<IdFilter<<Entity as Identifiable>::Id>>,
+            event_filter: FilterOpt<EventFilter>,
+        ) -> Self {
+            Self {
+                id_filter,
+                event_filter,
+            }
+        }
+    }
+
+    impl<Entity, EventFilter> Filter for ComplexEntityFilter<Entity, EventFilter>
+    where
+        Entity: IdTrait,
+        EventFilter: Filter<Item = Entity>,
+        for<'a> <Entity as Identifiable>::Id: IntoSchema + Deserialize<'a> + Serialize,
+    {
+        type Item = Entity;
+
+        fn filter(&self, entity: &Entity) -> bool {
+            self.id_filter.filter(entity.id()) && self.event_filter.filter(entity)
+        }
+    }
+}
+
+pub trait Filter {
+    type Item;
+
+    fn filter(&self, item: &Self::Item) -> bool;
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub enum FilterOpt<F: Filter> {
```

Comment:
we could use `Option<F>` instead

Reply:
I can implement `Filter` for `Option<F>`, but `Option` already has its own method `filter`. I don't think, that using fully quilified method calls in every match arm in this file will be good.
Also, I think `AcceptAll` is a little more readable than `None` for that purpose